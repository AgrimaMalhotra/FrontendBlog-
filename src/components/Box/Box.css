.box{
    width: 340px;
    height: 470px;
    border: #c6c6c6 solid 1px;
    margin: 3vh 2vw 3vh 2vw ;
    display: flex;
flex-direction: column;
}

.box-image{
    height: 65%;
}

.box-image img {
    width: 100%;
    height: 100%;
    object-fit: cover;
}

.box-content {
padding: 9px;
}

.box-content .date-duration {
    font-size: 10px;
    color: #c6c6c6;
    display: flex;
    justify-content: space-between;
}

.article-title p {
    margin-top: 6px;
    font-size: 20px;
    font-family: 'Lucida Sans', 'Lucida Sans Regular', 'Lucida Grande', 'Lucida Sans Unicode', Geneva, Verdana, sans-serif;
}

.article-description{
    height: 30%;
}
.article-description p {
    margin-top: 6px;
    text-align: justify;
    font-size: 15px;
    color: #696969;
    overflow: hidden;
white-space: nowrap;
text-overflow: ellipsis;
}

.like-bar {
    display: flex;
    justify-content: flex-end;
    flex-direction: column;
    padding-top: 10px;
    justify-self: flex-end;
}
.like-bar hr {
    margin-bottom: 5px;
}

.reaction{
    display: flex;
    justify-content: space-between;
}

.like-bar img {
    height: 20px;
    width: 20px;
}

.applause {
    display: flex;
}

.applause p {
    font-size: 14px;
    color: #696969;
    margin: 2px;
}
