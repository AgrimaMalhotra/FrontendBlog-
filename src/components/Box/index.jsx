import * as React from 'react';
import propTypes from 'prop-types';
import reaction from '../../assets/Icons/clapping.svg';
import heartBlack from '../../assets/Icons/heart-black.svg';
import heartRed from '../../assets/Icons/heart-red.svg';
import './Box.css';
import { UPDATE_BLOG_DATA } from '../../constants/apiEndPoints';
import makeRequest from '../../utils/makeRequest';
import { getFormattedDateFromUtcDate } from '../../utils/common';

function Box({ Data }) {
  const [isLiked, setIsLiked] = React.useState(Data.liked);
  const [clapCount, setClapCount] = React.useState(Data.claps);

  const handleLike = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(Data.id), {
        data: { liked: !isLiked },
      });
      setIsLiked(!isLiked);
    } catch (e) {
      return (
        <div className="backendError">
          <p>{e}</p>
        </div>
      );
    }
  };

  const handleClap = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(Data.id), {
        data: { claps: clapCount + 1 },
      });

      setClapCount(clapCount + 1);
    } catch (e) {
      return (
        <div className="backendError">
          <p>{e}</p>
        </div>
      );
    }
  };

  return (
    <div className="box" data-testid="blog-post">
      <div className="box-image">
        <img src={Data.image} alt="Blog Post Img" />
      </div>
      <div className="box-content">
        <div>
          <div className="date-duration">
            <p>{getFormattedDateFromUtcDate(Data.date)}</p>
            <p>{Data.reading_time} </p>
          </div>
          <div className="article-title">
            <p>{Data.title}</p>
          </div>
          <div className="article-description">
            <p>{Data.description}</p>
          </div>
        </div>
        <div className="like-bar">
          <hr />
          <div className="reaction">
            <div className="applause">
              <img
                src={reaction}
                alt="Clap"
                onClick={handleClap}
                onKeyDown={handleClap}
              />
              <p>{clapCount}</p>
            </div>
            <div className="liked" onClick={handleLike} onKeyDown={handleLike}>
              <img src={isLiked ? heartRed : heartBlack} alt="heart" />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

Box.propTypes = {
  Data: propTypes.shape({
    id: propTypes.number.isRequired,
    date: propTypes.string.isRequired,
    // eslint-disable-next-line camelcase
    reading_time: propTypes.string.isRequired,
    title: propTypes.string.isRequired,
    description: propTypes.string.isRequired,
    claps: propTypes.number.isRequired,
    liked: propTypes.bool.isRequired,
    image: propTypes.string.isRequired,
  }).isRequired,
};

export default Box;
