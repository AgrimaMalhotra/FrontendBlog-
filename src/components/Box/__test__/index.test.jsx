import { fireEvent, render, screen } from '@testing-library/react';
import * as React from 'react';
import Box from '..';

const Data = {
  date: '2nd Januray, 2018',
  readingTime: '2 mins',
  title: 'The future of abstract art and the culture ...',
  description:
    'Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...',
  claps: 10,
  liked: false,
  image: 'abstract.png',
};

describe('Box', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Box {...Data} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should change to red heart on clicking black heart', () => {
    render(<Box {...Data} />);
    expect(screen.queryByAltText('black heart').src).toContain(
      'heart-black.svg'
    );
    fireEvent.click(screen.queryByAltText('black heart'));
    expect(screen.queryByAltText('red heart').src).toContain('heart-red.svg');
  });
  it('should change to black heart on clicking red heart', () => {
    Data.liked = !Data.liked;
    render(<Box {...Data} />);
    expect(screen.queryByAltText('red heart').src).toContain('heart-red.svg');
    fireEvent.click(screen.queryByAltText('red heart'));
    expect(screen.queryByAltText('black heart').src).toContain(
      'heart-black.svg'
    );
  });

  it('should increase clap count on clicking clap button', () => {
    render(<Box {...Data} />);
    expect(screen.getByText(Data.claps.toString())).toBeTruthy();
    fireEvent.click(screen.queryByAltText('Clap'));
    expect(screen.getByText((Data.claps + 1).toString())).toBeTruthy();
  });
});
