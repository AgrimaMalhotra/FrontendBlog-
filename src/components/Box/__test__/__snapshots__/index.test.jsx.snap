// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box should render correctly 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <div
      class="box-image"
    >
      <img
        alt="abstract.png"
        src="abstract.png"
      />
    </div>
    <div
      class="box-content"
    >
      <div
        class="date-duration"
      >
        <p>
          2nd Januray, 2018
        </p>
        <p>
          2 mins 
        </p>
      </div>
      <div
        class="article-title"
      >
        <p>
          The future of abstract art and the culture ...
        </p>
      </div>
      <div
        class="article-description"
      >
        <p />
      </div>
      <hr />
      <div
        class="like-bar"
      >
        <div
          class="applause"
        >
          <img
            alt="Clap"
            src="clapping.svg"
          />
          <p>
            10
          </p>
        </div>
        <div
          class="heart"
        >
          <img
            alt="black heart"
            src="heart-black.svg"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
